Test glob support

pugnascotia opened the pull request:
Implement a new `testGlob` configuration option. This allows test files to be matched using extended glob patterns instead of regular expressions, as implemented by [micromatch](https://github.com/jonschlinkert/micromatch).

Per #926, globs are a frequently-used tool for matching files, and can be easier to use for filename matching. Other popular tools also use globs instead of regexs.

`testRegex` is still supported in this PR, but `testGlob` becomes the favoured option.

Existing tests have been modified where necessary to make them pass. New test cases have been added to `SearchSource-test.js` to check that glob matching works in addition to regex matching. A test has also been added to `normalize-test.js` to check that a deprecation warning is printed to the console.

pksjce commented on packages/jest-cli/src/__tests__/SearchSource-test.js:
```diff
@@ -50,10 +51,23 @@ describe('SearchSource', () => {
+    // micromatch doesn't support '..' through the globstar ('**') to avoid
+    // infinite recursion.
+
+    it('supports ../ paths and unix separators via textRegex', () => {
```
`testRegex` is not used in this test at all?

pugnascotia replied:
Good catch, this used to rely on the default, so I've changed it to explicitly set a regex.

develar commented:
Is it going to be merged?

pugnascotia commented:
The CI job is failing and I don't understand why - anyone have any suggestions?

cpojer commented:
There is some issue with eslint on travis, I'll figure it out. @pugnascotia thanks for working on this, it'll take me a bit longer to get to this PR as there are currently 20+ open requests and I'm still catching up with work at FB after my vacation.

pugnascotia commented:
No worries, just pleased to help out.

pugnascotia commented:
Any luck figuring out what was up with CI? Do you want me to merge up to `master` again?

cpojer commented:
Do you mind rebasing this?

pugnascotia commented:
I've rebased and squashed, and `npm test` passes, though the PR seems stuck on thinking there are still conflicts.

cpojer commented:
Are you sure you rebased against Jest's master branch? It doesn't seem like it.

pugnascotia commented:
What...I'm *sure* I fetched before rebasing. Weird. OK, should be there now.

thymikee commented:
@pugnascotia maybe you fetched from your fork's master, happens to me sometimes 😄

pugnascotia commented:
I've rebased (again) onto master, and this time I've actually fixed the problems that arose. In particular, in commit b4f65e1a I altered the order in which config keys are validated, so that keys that are still in force but are deprecated will trigger warnings.

Re: the `package.json` change, there isn't a dep change so it's right that `yarn.lock` hasn't been updated.

cpojer commented on packages/jest-util/src/messages.js:
```diff
@@ -9,10 +9,11 @@
 import type {AssertionResult, TestResult} from 'types/TestResult';
 
 const chalk = require('chalk');
+const mm = require('micromatch');
```
can we call this variable micromatch everywhere please?

pugnascotia replied:
Yep.

cpojer commented on packages/jest-runtime/src/shouldInstrument.js:
```diff
@@ -23,6 +23,11 @@ const shouldInstrument = (filename: Path, config: Config): boolean => {
   }
 
+  if (config.testGlob && config.testGlob.length
+    && micromatch.any(filename, config.testGlob)) {
```
the code style in Jest is:

```
if (
  config.testGlob &&
  config.testGlob.length &&
  micromatch.any(filename, config.testGlob)
) {
```

cpojer commented on packages/jest-runtime/src/index.js:
```diff
@@ -116,7 +114,6 @@ class Runtime {
     this._mocksPattern =
       config.mocksPattern ? new RegExp(config.mocksPattern) : null;
     this._shouldAutoMock = config.automock;
-    this._testRegex = new RegExp(replacePathSepForRegex(config.testRegex));
```
Nice, this seems to have been dead code.

cpojer commented on packages/jest-editor-support/src/Settings.js:
```diff
@@ -44,6 +47,10 @@ module.exports = class Settings extends EventEmitter {
+        '**/__tests__/**/*.js?(x)',
+        '**/?(*.)(spec|test).js?(x)',
+      ],
       testRegex: '(/__tests__/.*|\\.(test|spec))\\.jsx?$',
```
Why do we need both?

pugnascotia replied:
Um...we probably don't :-)

cpojer commented on packages/jest-config/src/normalize.js:
```diff
@@ -371,6 +376,11 @@ function normalize(config: InitialConfig, argv: Object = {}) {
+  if (config.testRegex && (!config.testGlob || config.testGlob.length === 0)) {
+    // Prevent default glob conflicting with any explicit `testRegex` value
+    newConfig.testGlob = [];
+  }
```
Seems like we should warn if both are given?

cpojer replied:
Wait, I don't really understand this. Why are you setting testGlob to an empty array if testGlob is empty?

pugnascotia replied:
Sorry, that's little messy, I'll tidy that up.

cpojer commented on .gitignore:
```diff
@@ -2,6 +2,8 @@
 *.swp
 *~
+.idea
+*.iml
```
do you mind removing these here and instead moving these into a global gitignore config for your computer?

cpojer commented on packages/jest-config/src/normalize.js:
```diff
@@ -319,6 +319,11 @@ function normalize(config: InitialConfig, argv: Object = {}) {
+
+      case 'testRegex':
+        value = config[key];
+        break;
```
why did you add this as a separate option in this switch case?

pugnascotia replied:
Gah, that's a leftover from an earlier implementation, when the deprecation warnings were generated here. Good catch.

cpojer commented:
Finally had some time to look at this. Am I understanding this correctly – it'll support both testGlob and testRegex and have the following behaviors:

* Use both testGlob and testRegex if they are specified.
* Print a deprecation warning for testRegex.

?

pugnascotia commented:
That was my thinking, yes. Since we can't (AFAIK) turn an arbitrary regex into a glob, I felt it was best to allow `testRegex` but print a warning, them remove it entirely in a later release.

pugnascotia commented:
That the branch rebased onto `facebook/master` again.

cpojer commented:
I feel like we shouldn't show the deprecation warning for testRegex yet. Instead, I'd recommend this:

* warn/throw if both testGlob and testRegex are given.
* don't warn about use of testRegex. We'll visibly deprecate it in two majors from now to reduce the churn.

Are there any better names than "testGlob"? Any way we can drop the "glob" from it? I don't like it, especially as we'll be ending up with a singular option in the future :)

thymikee commented:
maybe `testPattern`? But this may be misleading, as we use the word `pattern` in watch mode and it's for regex.

pugnascotia commented:
Rebased again and updated as per your comments. I liked `testMatch` so I went with that.

`perl -p -i -e 's/testGlob/testMatch/g' $(git grep -l testGlob)`

pugnascotia commented:
Rebased onto master again - any more feedback on this?

cpojer commented:
Thank you for working on this and for being so thorough. Sorry for the delay in reviews. This will be in Jest 19.
